hs-abci-sdk/src/Tendermint/SDK/BaseApp/Metrics.hs

```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Tendermint.SDK.BaseApp.Metrics where
+
+import           Data.String (IsString (..))
+import           Data.Text   (Text)
+import           Polysemy
+
+data CountName = CountName
+  { countName   :: Text
+  , countLabels :: [(Text, Text)]
+  } deriving (Eq, Ord)
+
+instance IsString CountName where
+  fromString s = CountName (fromString s) mempty
+
+data HistogramName = HistogramName
+  { histogramName    :: Text
+  , histogramLabels  :: [(Text, Text)]
+  , histogramBuckets :: [Double]
+  } deriving (Eq, Ord)
+
+instance IsString HistogramName where
+  fromString s = HistogramName (fromString s) mempty defaultBuckets
+    where defaultBuckets = [0.0001, 0.001, 0.01, 0.1, 0.25, 0.5, 0.75, 1, 10, 100]
+
+-- @NOTE: might need a NewCount/NewHistogram summand to clean up some usage quirks
+data Metrics m a where
+  -- | Increments the count of a specific message
+  IncCount :: CountName -> Metrics m ()
+  -- | Update a histogram
+  ObserveHistogram :: HistogramName -> Double -> Metrics m ()
```

Comment:
What is the target use case for this? Usually this would make sense to support a timer, but we already have a command for timer that doesn't need to use this to be interpreted

Reply:
Valid point.

I don't remember if i've noted it somewhere, but the usage of these summands are also to init a histogram, since the act of "adding a hist" to a prometheus registry doesn't allow to set the starting value. You can see this in the tests where `observeHistogram` is passed `0.0`.

Reply:
An alternative to this is changing `observeHistogram` to an init hist function and using `observeHistogram` as just an `IO` operation.